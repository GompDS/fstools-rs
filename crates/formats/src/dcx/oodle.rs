use std::{
    cmp::min,
    io::{Error, Read, Result},
};

use fstools_oodle_rt::{decoder::OodleDecoder, Compressor, Oodle, OODLELZ_BLOCK_LEN};

// SAFETY: `OodleLZDecoder` pointer is safe to use across several threads.
unsafe impl<R: Read> Sync for OodleReader<R> {}

// SAFETY: See above.
unsafe impl<R: Read> Send for OodleReader<R> {}

pub struct OodleReader<R: Read> {
    reader: R,

    /// The Oodle decoder instance created for this buffer.
    decoder: OodleDecoder,

    /// A sliding window of bytes decoded by the compressor, large enough to keep the past block in
    /// memory while the next block is decoded.
    decode_buffer: Box<[u8]>,

    /// The decoders position into the sliding window.
    decode_buffer_writer_pos: usize,

    /// The number of bytes that the consuming reader is lagging behind the decoder.
    decode_buffer_reader_lag: usize,

    /// Oodle requires at least [`OODLELZ_BLOCK_LEN`] bytes available in the input buffer, which
    /// the read buffer might not fit. Instead, we buffer to this intermediate buffer and treat
    /// it as a sliding window to ensure there are always `OODLELZ_BLOCK_LEN` bytes available
    /// to read.
    io_buffer: Box<[u8]>,

    /// The number of bytes available to read from [`io_buffer`], ending at [`io_buffer_pos`].
    io_buffer_writer_pos: usize,

    /// Current position within the IO buffer.
    io_buffer_reader_pos: usize,
}

impl<R: Read> OodleReader<R> {
    pub fn new(reader: R, uncompressed_size: u32) -> Option<Self> {
        let oodle = Oodle::current()?;
        let decoder = oodle.create_decoder(
            Compressor::OodleLZ_Compressor_Kraken,
            uncompressed_size as usize,
        )?;
        let decode_buffer = vec![0u8; 3 * 1024 * 1024].into_boxed_slice();
        let io_buffer = vec![0u8; OODLELZ_BLOCK_LEN as usize * 2].into_boxed_slice();

        Some(Self {
            // SAFETY: Pointer is validated to be non-null above.
            decoder,
            reader,
            decode_buffer,
            decode_buffer_writer_pos: 0,
            decode_buffer_reader_lag: 0,
            io_buffer,
            io_buffer_reader_pos: 0,
            io_buffer_writer_pos: 0,
        })
    }
}

impl<R: Read> Read for OodleReader<R> {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
        if buf.is_empty() {
            return Ok(0);
        }

        let dictionary_size = 2 * 1024 * 1024;
        let mut total_written = 0usize;

        while total_written < buf.len() {
            let wpos = self.decode_buffer_writer_pos;

            // Check if there's data to be written from the sliding window first
            if self.decode_buffer_reader_lag > 0 {
                let bytes_to_copy = min(self.decode_buffer_reader_lag, buf.len() - total_written);
                let start = wpos - self.decode_buffer_reader_lag;
                let end = start + bytes_to_copy;

                let src = &self.decode_buffer[start..end];
                let dest = &mut buf[total_written..total_written + bytes_to_copy];

                dest.copy_from_slice(src);

                self.decode_buffer_reader_lag -= bytes_to_copy;
                total_written += bytes_to_copy;

                continue;
            }

            self.io_buffer_writer_pos += self
                .reader
                .read(&mut self.io_buffer[self.io_buffer_writer_pos..])?;

            let data = &self.io_buffer[self.io_buffer_reader_pos..self.io_buffer_writer_pos];
            // Read and decode new data
            if data.is_empty() {
                break; // EOF reached
            }

            let out = self
                .decoder
                .decode_some(&mut self.decode_buffer, wpos, data)
                .ok_or(Error::other("Oodle decoder failed"))?;
            // SAFETY: OodleLZ_DecodeSome_out is zero initialised by default.

            let decoded_bytes = out.decodedCount as usize;
            let consumed_bytes = out.compBufUsed as usize;

            self.io_buffer_reader_pos += consumed_bytes;

            if decoded_bytes > 0 {
                let bytes_to_copy = min(decoded_bytes, buf.len() - total_written);
                let dest = &mut buf[total_written..total_written + bytes_to_copy];
                let src = &self.decode_buffer[wpos..wpos + bytes_to_copy];

                dest.copy_from_slice(src);

                self.decode_buffer_writer_pos += decoded_bytes;
                self.decode_buffer_reader_lag = decoded_bytes - bytes_to_copy;
                total_written += bytes_to_copy;
            } else {
                // Nothing more to decode.
                if out.curQuantumCompLen == 0 {
                    return Ok(total_written);
                }

                let remaining = self.io_buffer_writer_pos - self.io_buffer_reader_pos;

                self.io_buffer.rotate_left(self.io_buffer_reader_pos);
                self.io_buffer_reader_pos = 0;
                self.io_buffer_writer_pos = remaining;
            }

            // Manage sliding window
            if self.decode_buffer_writer_pos + OODLELZ_BLOCK_LEN as usize > self.decode_buffer.len()
            {
                self.decode_buffer.copy_within(
                    self.decode_buffer_writer_pos - dictionary_size..self.decode_buffer_writer_pos,
                    0,
                );

                self.decode_buffer_writer_pos = dictionary_size;
            }
        }

        Ok(total_written)
    }
}
