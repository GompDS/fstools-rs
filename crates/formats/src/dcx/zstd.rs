use std::io::{self, Read};

/// Trivial wrapper around a [`zstd::Decoder<BufReader<R>>`].
pub struct ZstdDecoder<R: Read>(zstd::Decoder<'static, io::BufReader<R>>);

impl<R: Read> ZstdDecoder<R> {
    pub fn new(reader: R) -> io::Result<Self> {
        Ok(Self(zstd::Decoder::new(reader)?))
    }
}

impl<R: Read> Read for ZstdDecoder<R> {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        self.0.read(buf)
    }
}
