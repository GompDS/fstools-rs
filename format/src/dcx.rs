use std::{io, mem};

use byteorder::{ReadBytesExt, BE};

use thiserror::Error;

use crate::io_ext::ReadFormatsExt;

#[derive(Debug, Error)]
pub enum DCXError {
    #[error("Could not copy bytes {0}")]
    Io(#[from] io::Error),

    #[error("Got error from oodle decompression: {0}")]
    Decompress(u32),
}

#[derive(Debug)]
pub struct DCX {
    pub unk04: u32,
    pub dcs_offset: u32,
    pub dcp_offset: u32,
    pub unk10: u32,
    pub unk14: u32,
    pub dcs: u32,
    pub uncompressed_size: u32,
    pub compressed_size: u32,
    pub dcp: u32,
    pub format: u32,
    pub unk2c: u32,
    pub compression_level: u8,
    pub unk31: u8,
    pub unk32: u8,
    pub unk33: u8,
    pub unk34: u32,
    pub unk38: u32,
    pub unk3c: u32,
    pub unk40: u32,
    pub dca: u32,
    pub dca_size: u32,
    pub decompressed: Vec<u8>,
}

impl DCX {
    pub fn from_reader(r: &mut impl io::Read) -> Result<Self, DCXError> {
        r.read_magic(b"DCX\0")?;

        let unk04 = r.read_u32::<BE>()?;
        let dcs_offset = r.read_u32::<BE>()?;
        let dcp_offset = r.read_u32::<BE>()?;
        let unk10 = r.read_u32::<BE>()?;
        let unk14 = r.read_u32::<BE>()?;
        let dcs = r.read_u32::<BE>()?;
        let uncompressed_size = r.read_u32::<BE>()?;
        let compressed_size = r.read_u32::<BE>()?;
        let dcp = r.read_u32::<BE>()?;
        let format = r.read_u32::<BE>()?;
        assert!(format == 0x4b52414b, "Format was not KRAKEN");

        let unk2c = r.read_u32::<BE>()?;
        let compression_level = r.read_u8()?;
        let unk31 = r.read_u8()?;
        let unk32 = r.read_u8()?;
        let unk33 = r.read_u8()?;
        let unk34 = r.read_u32::<BE>()?;
        let unk38 = r.read_u32::<BE>()?;
        let unk3c = r.read_u32::<BE>()?;
        let unk40 = r.read_u32::<BE>()?;
        let dca = r.read_u32::<BE>()?;
        let dca_size = r.read_u32::<BE>()?;

        let mut compressed = vec![0x0u8; compressed_size as usize];
        r.read_exact(&mut compressed)?;

        let mut decompressed = vec![0x0u8; uncompressed_size as usize];

        oodle_safe::decompress(
            &compressed,
            &mut decompressed,
            None,
            None,
            None,
            None,
        ).map_err(DCXError::Decompress)?;

        Ok(Self {
            unk04,
            dcs_offset,
            dcp_offset,
            unk10,
            unk14,
            dcs,
            uncompressed_size,
            compressed_size,
            dcp,
            format,
            unk2c,
            compression_level,
            unk31,
            unk32,
            unk33,
            unk34,
            unk38,
            unk3c,
            unk40,
            dca,
            dca_size,
            decompressed,
        })
    }

    pub fn has_magic(r: &mut (impl io::Read + io::Seek)) -> Result<bool, io::Error> {
        // Read magic and check if it's DCX
        let result = r.read_u32::<BE>()? == 0x44435800;

        // Seek backwards to before the magic
        r.seek(io::SeekFrom::Current(-(mem::size_of::<u32>() as i64)))?;

        Ok(result)
    }
}
