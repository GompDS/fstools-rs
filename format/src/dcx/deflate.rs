use std::io::{self, Read};
use flate2::{read::ZlibDecoder, Compression};

pub struct DcxDecoderDeflate<'a>(ZlibDecoder<&'a [u8]>);

impl<'a> DcxDecoderDeflate<'a> {
    pub fn from_buffer(buf: &'a [u8]) -> Self {
        Self(ZlibDecoder::new(buf))
    }
}

impl<'a> Read for DcxDecoderDeflate<'a> {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        self.0.read(buf)
    }
}
