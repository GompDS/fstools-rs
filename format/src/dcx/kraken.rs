use std::{
    alloc::alloc,
    cmp::min,
    io::{BufRead, BufReader, Error, Read, Result, Take},
    ptr::null_mut,
};

use oodle_sys::{
    OodleLZDecoder, OodleLZDecoder_Create, OodleLZDecoder_DecodeSome, OodleLZDecoder_Destroy,
    OodleLZDecoder_MemorySizeNeeded, OodleLZ_CheckCRC_OodleLZ_CheckCRC_No,
    OodleLZ_CheckCRC_OodleLZ_CheckCRC_Yes, OodleLZ_Compressor_OodleLZ_Compressor_Invalid,
    OodleLZ_Compressor_OodleLZ_Compressor_Kraken, OodleLZ_DecodeSome_Out,
    OodleLZ_Decode_ThreadPhase_OodleLZ_Decode_Unthreaded, OodleLZ_FuzzSafe_OodleLZ_FuzzSafe_No,
    OodleLZ_FuzzSafe_OodleLZ_FuzzSafe_Yes, OodleLZ_Verbosity_OodleLZ_Verbosity_Lots,
    OodleLZ_Verbosity_OodleLZ_Verbosity_None, OODLELZ_BLOCK_LEN,
};

pub struct DcxDecoderKraken<R: Read> {
    reader: R,

    /// The total size of the raw data expected to be read from the underlying stream.
    uncompressed_size: u32,

    /// The Oodle decoder instance created for this buffer.
    decoder: *mut OodleLZDecoder,

    /// A sliding window of bytes decoded by the compressor, large enough to keep the past block in
    /// memory while the next block is decoded.
    decode_buffer: Box<[u8]>,

    /// The decoders position into the sliding window.
    decode_buffer_pos: usize,

    /// The number of bytes that the consuming reader is lagging behind the decoder.
    decode_buffer_lag: usize,
    io_buffer: Box<[u8]>,

    /// The number of bytes available to read from [io_buffer], ending at [io_buffer_pos].
    io_buffer_end: usize,

    /// Current position within the IO buffer.
    io_buffer_pos: usize,
}

impl<R: Read> DcxDecoderKraken<R> {
    // TODO: fix vfs reader so it isn't producing padding
    pub fn new(reader: R, uncompressed_size: u32) -> Self {
        let compressor = OodleLZ_Compressor_OodleLZ_Compressor_Invalid;
        let decoder = unsafe {
            OodleLZDecoder_Create(compressor, uncompressed_size as i64, null_mut(), 0isize)
        };

        if decoder.is_null() {
            panic!("return error here: failed to create decoder, check oodle error");
        }

        let decode_buffer = vec![0u8; 3 * 1024 * 1024].into_boxed_slice();
        let io_buffer = vec![0u8; OODLELZ_BLOCK_LEN as usize * 2].into_boxed_slice();

        Self {
            decoder,
            reader,
            decode_buffer,
            decode_buffer_pos: 0,
            decode_buffer_lag: 0,
            io_buffer,
            io_buffer_pos: 0,
            io_buffer_end: 0,
            uncompressed_size,
        }
    }
}

impl<R: Read> Read for DcxDecoderKraken<R> {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
        if buf.is_empty() {
            return Ok(0);
        }

        let dictionary_size = 2 * 1024 * 1024;
        let mut total_written = 0usize;

        while total_written < buf.len() {
            let wpos = self.decode_buffer_pos;

            // Check if there's data to be written from the sliding window first
            if self.decode_buffer_lag > 0 {
                let bytes_to_copy = min(self.decode_buffer_lag, buf.len() - total_written);
                let start = wpos - self.decode_buffer_lag;
                let end = start + bytes_to_copy;

                let src = &self.decode_buffer[start..end];
                let dest = &mut buf[total_written..total_written + bytes_to_copy];

                dest.copy_from_slice(src);

                self.decode_buffer_lag -= bytes_to_copy;
                total_written += bytes_to_copy;

                continue;
            }

            self.io_buffer_end += self
                .reader
                .read(&mut self.io_buffer[self.io_buffer_end..])?;

            let data = &self.io_buffer[self.io_buffer_pos..self.io_buffer_end];
            // Read and decode new data
            if data.is_empty() {
                break; // EOF reached
            }

            let mut out: OodleLZ_DecodeSome_Out = unsafe { std::mem::zeroed() };
            let result = unsafe {
                // EXTREMELY unlikely, however unsound otherwise.
                let input_data_len = isize::try_from(data.len()).unwrap_or(isize::MAX);

                // SAFETY:
                // - Signedness conversions of offsets are all valid, given that
                //   `sliding_window.len() <= i32::MAX` and `self.uncompressed_size < isize::MAX`.
                // - Consumed `input_data_len` is caped at i32::MAX
                let decode_buffer_avail = (self.decode_buffer.len() - wpos);
                OodleLZDecoder_DecodeSome(
                    self.decoder,
                    &mut out as *mut _,
                    self.decode_buffer.as_mut_ptr() as *mut _,
                    wpos as isize,
                    self.uncompressed_size as isize,
                    decode_buffer_avail as isize,
                    data.as_ptr() as *const _,
                    input_data_len,
                    OodleLZ_FuzzSafe_OodleLZ_FuzzSafe_Yes,
                    OodleLZ_CheckCRC_OodleLZ_CheckCRC_Yes,
                    OodleLZ_Verbosity_OodleLZ_Verbosity_None,
                    OodleLZ_Decode_ThreadPhase_OodleLZ_Decode_Unthreaded,
                )
            };

            if result == 0 {
                return Err(Error::other("Oodle decoder failed"));
            }

            let decoded_bytes = out.decodedCount as usize;
            let consumed_bytes = out.compBufUsed as usize;

            self.io_buffer_pos += consumed_bytes;

            if decoded_bytes > 0 {
                let bytes_to_copy = min(decoded_bytes, buf.len() - total_written);
                let dest = &mut buf[total_written..total_written + bytes_to_copy];
                let src = &self.decode_buffer[wpos..wpos + bytes_to_copy];

                dest.copy_from_slice(src);

                self.decode_buffer_pos += decoded_bytes;
                self.decode_buffer_lag = decoded_bytes - bytes_to_copy;
                total_written += bytes_to_copy;
            } else {
                // Nothing more to decode.
                if out.curQuantumCompLen == 0 {
                    return Ok(0);
                }

                let remaining = self.io_buffer_end - self.io_buffer_pos;

                self.io_buffer.rotate_left(self.io_buffer_pos);
                self.io_buffer_pos = 0;
                self.io_buffer_end = remaining;
            }

            // Manage sliding window
            if self.decode_buffer_pos + OODLELZ_BLOCK_LEN as usize > self.decode_buffer.len() {
                self.decode_buffer.copy_within(
                    self.decode_buffer_pos - dictionary_size..self.decode_buffer_pos,
                    0,
                );

                self.decode_buffer_pos = dictionary_size;
            }
        }

        Ok(total_written)
    }
}

impl<R: Read> Drop for DcxDecoderKraken<R> {
    fn drop(&mut self) {
        unsafe { OodleLZDecoder_Destroy(self.decoder) }
    }
}
