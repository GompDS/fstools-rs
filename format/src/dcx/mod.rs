use std::{io, io::Read};

use byteorder::BE;
use thiserror::Error;
use zerocopy::{AsBytes, FromBytes, FromZeroes, Ref, U32};

use self::{deflate::DcxDecoderDeflate, kraken::DcxDecoderKraken};

pub mod deflate;
pub mod kraken;

const MAGIC_DCX: u32 = 0x44435800;
const MAGIC_ALGORITHM_KRAKEN: &[u8; 4] = b"KRAK";
const MAGIC_ALGORITHM_DEFLATE: &[u8; 4] = b"DFLT";

#[derive(Debug, Error)]
pub enum DcxError {
    #[error("Could not copy bytes {0}")]
    Io(#[from] io::Error),

    #[error("Unrecognized DCX compression algorithm: {0:x?}")]
    UnknownAlgorithm([u8; 4]),

    #[error("Could not properly parse DCX file")]
    ParserError,
}

#[derive(Debug, Error)]
pub enum DecompressionError {
    #[error("Oodle error code: {0}")]
    Oodle(u32),

    #[error("Zlib error")]
    Zlib,
}

#[derive(FromZeroes, FromBytes)]
#[repr(C)]
#[allow(unused)]
pub struct Dcx<'a> {
    bytes: &'a [u8],

    header: &'a Header,
    sizes: &'a Sizes,
    compression_parameters: &'a CompressionParameters,
    compressed: &'a [u8],
}

impl<'a> Dcx<'a> {
    // TODO: add magic validation
    pub fn parse(bytes: &'a [u8]) -> Option<Self> {
        let (header, next) = Ref::<_, Header>::new_from_prefix(bytes)?;
        let (sizes, next) = Ref::<_, Sizes>::new_from_prefix(next)?;
        let (compression_parameters, next) =
            Ref::<_, CompressionParameters>::new_from_prefix(next)?;
        let (additional, rest) = Ref::<_, Additional>::new_from_prefix(next)?;

        Some(Self {
            bytes,
            header: header.into_ref(),
            sizes: sizes.into_ref(),
            compression_parameters: compression_parameters.into_ref(),
            compressed: rest,
        })
    }

    pub fn create_decoder(&self) -> Result<DcxContentDecoder, DcxError> {
        let algorithm = &self.compression_parameters.algorithm;
        let decoder = match algorithm {
            MAGIC_ALGORITHM_KRAKEN => Decoder::Kraken(DcxDecoderKraken::from_buffer(
                self.compressed,
                self.sizes.uncompressed_size,
            )),
            MAGIC_ALGORITHM_DEFLATE => {
                Decoder::Deflate(DcxDecoderDeflate::from_buffer(self.compressed))
            }
            _ => return Err(DcxError::UnknownAlgorithm(algorithm.to_owned())),
        };

        Ok(DcxContentDecoder {
            uncompressed_size: self.sizes.uncompressed_size,
            compressed: &self.compressed,
            decoder,
        })
    }

    pub fn has_magic(buf: &[u8]) -> bool {
        match U32::<BE>::ref_from_prefix(buf) {
            Some(v) => v.get() == MAGIC_DCX,
            None => false,
        }
    }
}

impl<'a> std::fmt::Debug for Dcx<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("DCX")
            .field("header", self.header)
            .field("sizes", self.sizes)
            .field("compression_parameters", self.compression_parameters)
            .finish()
    }
}

pub enum Decoder<'a> {
    Kraken(DcxDecoderKraken<'a>),
    Deflate(DcxDecoderDeflate<'a>),
}

pub struct DcxContentDecoder<'a> {
    /// Reference to the compressed bytes.
    compressed: &'a [u8],

    /// Size of the contents once decompressed.
    uncompressed_size: U32<BE>,

    decoder: Decoder<'a>,
}

impl<'a> DcxContentDecoder<'a> {
    pub fn hint_size(&self) -> usize {
        self.uncompressed_size.get() as usize
    }
}

impl<'a> Read for DcxContentDecoder<'a> {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        match &mut self.decoder {
            Decoder::Kraken(d) => d.read(buf),
            Decoder::Deflate(d) => d.read(buf),
        }
    }
}

#[derive(FromZeroes, FromBytes, Debug)]
#[repr(C)]
#[allow(unused)]
/// The DCX chunk. Describes the layout of the container.
struct Header {
    chunk_magic: [u8; 4],

    /// Overal Dcx file version
    version: U32<BE>,

    /// Offset to the DCS chunk
    sizes_offset: U32<BE>,

    /// Offset to the DCP chunk
    params_offset: U32<BE>,

    /// Offset to the DCA chunk
    data_info_offset: U32<BE>,

    /// Offset to the compressed data
    data_offset: U32<BE>,
}

#[derive(FromZeroes, FromBytes, Debug)]
#[repr(C)]
#[allow(unused)]
/// The DCS Chunk. Describes the sizes before and after compression.
struct Sizes {
    chunk_magic: [u8; 4],
    /// Size of the data when decompressed, can be used for reserving vector
    /// capacity.
    uncompressed_size: U32<BE>,

    /// Size of the data in its compressed form.
    compressed_size: U32<BE>,
}

#[derive(FromZeroes, FromBytes, Debug)]
#[repr(C)]
#[allow(unused)]
/// The DCP chunk. Describes parameters used for compression/decompression.
struct CompressionParameters {
    chunk_magic: [u8; 4],

    /// Either KRAK, DFLT or EDGE
    algorithm: [u8; 4],

    /// Seems to the size of the current DCP chunk including magic and algo
    chunk_size: U32<BE>,

    /// Arbitrary bytes describing the parameter chunk
    /// TODO make this of dynamic size using the chunksize
    settings: [u8; 20],
}

#[derive(FromZeroes, FromBytes, Debug)]
#[repr(C)]
#[allow(unused)]
/// The DCA chunk. Describes ???
struct Additional {
    chunk_magic: [u8; 4],

    /// Seems to the size of the current DCA chunk including magic
    chunk_size: U32<BE>,
}
