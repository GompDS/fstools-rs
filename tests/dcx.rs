use std::{
    collections::HashSet,
    error::Error,
    ffi::OsStr,
    io::{self, Read},
    path::{Path, PathBuf},
    sync::Arc,
};

use fstools::{dvdbnd, formats::dcx::DcxHeader, prelude::*};
use fstools_dvdbnd::GameType::EldenRing;
use fstools_elden_ring_support::decrypt_regulation;
use fstools_formats::dcx::DcxError;
use insta::assert_snapshot;
use libtest_mimic::{Arguments, Failed, Trial};

fn main() -> Result<(), Box<dyn Error>> {
    let args = Arguments::from_args();
    let er_path = PathBuf::from(std::env::var("ER_PATH").expect("er_path"));
    let reg_path = er_path.join("regulation.bin");
    let keys_path = PathBuf::from(std::env::var("ER_KEYS_PATH").expect("er_keys_path"));
    let vfs = Arc::new(DvdBnd::create_from_game(
        EldenRing,
        er_path,
        FileKeyProvider::new(keys_path),
    )?);

    let lines = DvdBnd::dictionary_from_game(EldenRing)
        .filter(|line| line.extension() == Some(OsStr::new("dcx")))
        .collect::<HashSet<_>>();

    let mut tests = vec![];

    assert_snapshot!(lines.len());

    for line in lines {
        let vfs = vfs.clone();
        let test = Trial::test(line.to_string_lossy().to_string(), move || {
            check_file(vfs.clone(), &line)
        })
        .with_kind("dcx");

        tests.push(test);
    }

    // Test against the regulation DCX (ZSTD encoded since 1.12)
    tests.push(Trial::test("regulation.bin", move || {
        check_regulation(&reg_path)
    }));

    libtest_mimic::run(&args, tests).exit();
}

pub fn check_regulation(path: &Path) -> Result<(), Failed> {
    let regulation_bytes = std::fs::read(path.join("regulation.bin"))?;
    let dcx_bytes = decrypt_regulation(&mut regulation_bytes.as_slice())?;
    check_dcx(io::Cursor::new(dcx_bytes))
}

pub fn check_file(vfs: Arc<DvdBnd>, file: &Path) -> Result<(), Failed> {
    let file = match vfs.open(file.to_string_lossy().as_ref()) {
        Ok(file) => file,
        Err(_) => {
            return Ok(());
        }
    };
    check_dcx(file)
}

pub fn check_dcx(reader: impl Read) -> Result<(), Failed> {
    let (_, mut decoder) = match DcxHeader::read(reader) {
        Ok(details) => details,
        Err(DcxError::UnknownAlgorithm(_)) => return Ok(()),
        Err(_) => return Err("failed to parse DCX header".into()),
    };

    std::io::copy(&mut decoder, &mut std::io::sink())?;

    Ok(())
}
