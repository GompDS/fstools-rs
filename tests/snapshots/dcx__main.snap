---
source: tests/dcx.rs
expression: lines.len()
---
69821
